### PR TITLE
SALTO-5866 confluence e2e

### PR DESCRIPTION
**Pull request by omrifarkash**
---

- add e2e_utils to adapter-components

---
_Release Notes_: 
Confluence Adapter_:
- e2e test

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/confluence-adapter/e2e_test/mock_elements.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      labels: {
+        start: 0,
+        limit: 200,
+        size: 0,
+      },
+    },
+    type: 'global',
+    status: 'current',
+  },
+  [PAGE_TYPE_NAME]: {
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/mock_elements.ts:41]: is this type usually that simple ?

File: packages/confluence-adapter/e2e_test/mock_elements.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Values } from '@salto-io/adapter-api'
+import { SPACE_TYPE_NAME, PAGE_TYPE_NAME, TEMPLATE_TYPE_NAME } from '../src/constants'
+
+export const mockDefaultValues: Record<string, Values> = {
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/mock_elements.ts:19]: nit: I think it's worth adding the instances we test something with reference / parent annotation, so that the E2E would cover the reference resolving as well.

File: packages/confluence-adapter/e2e_test/adapter.test.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        .map(res => res.appliedChanges)
+        .flat()
+        .map(change => getChangeData(change)) as InstanceElement[]
+
+      deployInstances.forEach(deployedInstance => {
+        const { typeName } = deployedInstance.elemID
+        const instance = elements
+          .filter(isInstanceElement)
+          .filter(inst => inst.elemID.typeName === typeName)
+          .find(inst => inst.value[nameFieldPerType[typeName]] === deployedInstance.value[nameFieldPerType[typeName]])
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/adapter.test.ts:267]: why are you comparing `nameFieldPerType` and not elemID?

Also, probably worth creating `_.keyBy(elements, e => e.elemID.getFullName())` once instead of filtering the list for each `deployedInstance`

File: packages/confluence-adapter/e2e_test/adapter.test.ts [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+const TEST_PREFIX = 'Test'
+
+const nameFieldPerType: Record<string, string> = {
+  [SPACE_TYPE_NAME]: 'key',
+  [PAGE_TYPE_NAME]: 'title',
+  [TEMPLATE_TYPE_NAME]: 'name',
+}
+
+const fieldsToComparePerType: Record<string, string[]> = {
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/adapter.test.ts:75]: Why are we limiting the compare to specific fields ?

File: packages/confluence-adapter/e2e_test/adapter.test.ts [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+          if (planElement !== undefined) {
+            applyDetailedChanges(planElement, detailedCompare(planElement, updatedElement))
+          }
+        })
+      return deployResult
+    })
+    .toArray()
+}
+
+const checkNameField = (instance: InstanceElement): boolean => {
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/adapter.test.ts:166]: why is this needed ?
Also I think we shouldn't have `nameFieldPerType` cause it's a duplication for `elemID` definitions, and changing the fields used for elemID would break this test

File: packages/confluence-adapter/e2e_test/adapter.test.ts [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+  return [toChange({ after: spaceInstance }), toChange({ after: pageInstance }), toChange({ after: templateInstance })]
+}
+
+const nullProgressReporter: ProgressReporter = {
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  reportProgress: () => {},
+}
+
+const deployChanges = async (adapterAttr: Reals, changes: Change[]): Promise<DeployResult[]> => {
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/adapter.test.ts:142]: is this just a copy paste from google-workspace `deployChanges`? if so lets create e2e utils and make both e2e's use this

File: packages/confluence-adapter/e2e_test/adapter.test.ts [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+const checkNameField = (instance: InstanceElement): boolean => {
+  const { typeName } = instance.elemID
+  const nameField = instance.value[nameFieldPerType[typeName]]
+  return typeof nameField === 'string' && nameField.startsWith(TEST_PREFIX)
+}
+
+const getChangesForInitialCleanup = (elements: InstanceElement[]): Change<InstanceElement>[] =>
+  elements.filter(checkNameField).map(instance => toChange({ before: instance }))
+
+const deployCleanup = async (adapterAttr: Reals, elements: InstanceElement[]): Promise<void> => {
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/adapter.test.ts:175]: relevant for here too, let's not duplicate?

File: packages/confluence-adapter/e2e_test/adapter.test.ts [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    types,
+    uniqueValue: createUniqueValue(PAGE_TYPE_NAME),
+  })
+
+  const templateInstance = createInstance({
+    typeName: TEMPLATE_TYPE_NAME,
+    types,
+    uniqueValue: createUniqueValue(TEMPLATE_TYPE_NAME),
+  })
+  spaceInstance.value.name = createUniqueValue(SPACE_TYPE_NAME)
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/adapter.test.ts:129]: nit: you can add `createInstance` with `valuesOverride` param and then override default values provided by `mockDefaultValues`

File: packages/confluence-adapter/e2e_test/adapter.test.ts [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+  const spaceRef = new ReferenceExpression(spaceInstance.elemID, spaceInstance)
+  pageInstance.value.spaceId = spaceRef
+  templateInstance.annotate({ [CORE_ANNOTATIONS.PARENT]: [spaceRef] })
```
Comment (shir-reifenberg) [inline packages/confluence-adapter/e2e_test/adapter.test.ts:132]: I now see you did included a reference :) please generalize `createInstance` to have a parent param for easier extension of this test

**Review by shir-reifenberg — COMMENTED** [review]
Looks good 😎  Added some questions and comments
As is seems like there's a pretty similar structure to this test and google workspace test, I think we should consider consolidate what ever is possible for faster creation of next tests

File: packages/confluence-adapter/e2e_test/mock_elements.ts (hunk 1 above)
Comment (omrifarkash) [reply packages/confluence-adapter/e2e_test/mock_elements.ts:41]: Yes

File: packages/confluence-adapter/e2e_test/mock_elements.ts (hunk 2 above)
Comment (omrifarkash) [reply packages/confluence-adapter/e2e_test/mock_elements.ts:19]: I am doing it on the test

File: packages/confluence-adapter/e2e_test/adapter.test.ts (hunk 3 above)
Comment (omrifarkash) [reply packages/confluence-adapter/e2e_test/adapter.test.ts:267]: - Because I created the instances without a real computeElemId func so the elemIds won't be the same.
- 👍

File: packages/confluence-adapter/e2e_test/adapter.test.ts (hunk 4 above)
Comment (omrifarkash) [reply packages/confluence-adapter/e2e_test/adapter.test.ts:75]: Because there are fields we expect to be different. I can work with black list if you prefer

File: packages/confluence-adapter/e2e_test/adapter.test.ts (hunk 5 above)
Comment (omrifarkash) [reply packages/confluence-adapter/e2e_test/adapter.test.ts:166]: These are not necessarily equal to elemId parts. but fields that must be unique

For example, key on space instance

File: packages/confluence-adapter/e2e_test/adapter.test.ts (hunk 6 above)
Comment (omrifarkash) [reply packages/confluence-adapter/e2e_test/adapter.test.ts:142]: It is as most of the code here. I think Neta wants to make something generic after we will see more adapters. There is a different task for it

File: packages/confluence-adapter/e2e_test/adapter.test.ts (hunk 4 above)
Comment (shir-reifenberg) [reply packages/confluence-adapter/e2e_test/adapter.test.ts:75]: I think that if something changes in the APIs it would be difficult to detect it this way.
in what cases we expect fields to be different?

File: packages/confluence-adapter/e2e_test/adapter.test.ts (hunk 3 above)
Comment (shir-reifenberg) [reply packages/confluence-adapter/e2e_test/adapter.test.ts]: not sure I understand, let's discuss IRL ?

File: packages/confluence-adapter/e2e_test/adapter.test.ts (hunk 6 above)
Comment (shir-reifenberg) [reply packages/confluence-adapter/e2e_test/adapter.test.ts:155]: whatever you and @netama decide, but IMO as there are at least 3 more adapters in progress now, I think consolidating the code some of this code is not a lot of effort and will save us times in the next E2Es

File: packages/adapter-components/src/e2e_utils.ts
Comment (shir-reifenberg) [inline packages/adapter-components/src/e2e_utils.ts:1]: maybe just for consistency with other packages put this in `packages/adapter-components/e2e_test/utils.ts` ?

**Review by shir-reifenberg — APPROVED** [review]
Looks great! 👑 
added mainly nits but nothing major
